graphistry/collapse.py

```diff
@@ -0,0 +1,532 @@
+
+    g._edges = edf
+    return g
+
+
+def has_property(g, ref_node, attribute, column: str) -> bool:
+    """
+        Checks if ref_node is in node dataframe in column with attribute
+
+    -------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param ref_node: `node` to check if it as `attribute` in `column`
+    :returns bool"""
+    ndf, edf, src, dst, node = unpack(g)
+    ref_node = unwrap_key(ref_node)
+    return ref_node in ndf[ndf[column] == attribute][node].values
+
+
+def check_default_columns_present_and_coerce_to_string(g):
+    """
+        Helper to set COLLAPSE columns to nodes and edges dataframe, while converting src, dst, node to dtype(str)
+
+    -------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :returns graphistry instance
+    """
+    ndf, edf, src, dst, node = unpack(g)
+    if COLLAPSE_NODE not in ndf.columns:
+        ndf[COLLAPSE_NODE] = DEFAULT_VAL
+        ndf[node] = ndf[node].astype(str)
+        print(f'Converted ndf to type({type(ndf[node].values[0])})')
+        g._nodes = ndf
+    if COLLAPSE_SRC not in edf.columns:
+        edf[COLLAPSE_SRC] = DEFAULT_VAL
+        edf[COLLAPSE_DST] = DEFAULT_VAL
+        edf[src] = edf[src].astype(str)
+        edf[dst] = edf[dst].astype(str)
+        print(f'Converted edf to type({type(edf[src].values[0])})')
+        g._edges = edf
+    return g
+
+
+def collapse(g, child, parent, attribute, column: str, seen: dict):
+    """
+        Basically candy crush over graph properties in a topology aware manner
+
+        Checks to see if child node has desired property from parent, we will need to check if
+        (start_node=parent: has_attribute , children nodes: has_attribute) by case
+        (T, T), (F, T), (T, F) and (F, F),
+        we start recursive collapse (or not) on the children, reassigning nodes and edges.
+
+        if (T, T), append children nodes to start_node, re-assign the name of the node, and update the edge table with new name,
+
+        if (F, T) start k-(potentially new) super nodes, with k the number of children of start_node.
+                Start node keeps k outgoing edges.
+
+        if (T, F) it is the end of the cluster, and we keep new node as is; keep going
+
+        if (F, F); keep going
+
+    --------------------------------------------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param child: child node to start traversal, for first traversal, set child=parent or vice versa.
+    :param parent: parent node to start traversal, in main call, this is set to child.
+    :param attribute: attribute to collapse by
+    :param column: column in nodes dataframe to collapse over.
+    :returns graphistry instance with collapsed nodes.
+
+    """
+    g = check_default_columns_present_and_coerce_to_string(g)
```

Comment:
this gets called on every recursion, is that right? or just should be at entrypoint?

Reply:
Every recursion, but lightweight bools overall. I can move it out of scope, and change `collapse` -> `_collapse`